cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/KafkaClusterOperation.java

```diff
@@ -0,0 +1,263 @@
+            result.add(futureHeadlessService);
+
+            Future<Void> futureStatefulSet = Future.future();
+            CreateOperation.createStatefulSet(vertx, k8s.getKubernetesClient()).create(kafka.generateStatefulSet(k8s.isOpenShift()), futureStatefulSet.completer());
+            result.add(futureStatefulSet);
+
+            return result;
+        }
+    };
+
+    private final Op<KafkaCluster> delete = new Op<KafkaCluster>() {
+        @Override
+        public List<Future> futures(K8SUtils k8s, String namespace, KafkaCluster kafka) {
+            boolean deleteClaims = kafka.getStorage().type() == Storage.StorageType.PERSISTENT_CLAIM
+                    && kafka.getStorage().isDeleteClaim();
+            List<Future> result = new ArrayList<>(4 + (deleteClaims ? kafka.getReplicas() : 0));
+
+            if (kafka.isMetricsEnabled()) {
+                Future<Void> futureConfigMap = Future.future();
+                DeleteOperation.deleteConfigMap(vertx, k8s.getKubernetesClient()).delete(namespace, kafka.getMetricsConfigName(), futureConfigMap.completer());
+                result.add(futureConfigMap);
+            }
+
+            Future<Void> futureService = Future.future();
+            DeleteOperation.deleteService(vertx, k8s.getKubernetesClient()).delete(namespace, kafka.getName(), futureService.completer());
+            result.add(futureService);
+
+            Future<Void> futureHeadlessService = Future.future();
+            DeleteOperation.deleteService(vertx, k8s.getKubernetesClient()).delete(namespace, kafka.getHeadlessName(), futureHeadlessService.completer());
+            result.add(futureHeadlessService);
+
+            Future<Void> futureStatefulSet = Future.future();
+            DeleteOperation.deleteStatefulSet(vertx, k8s.getKubernetesClient()).delete(namespace, kafka.getName(), futureStatefulSet.completer());
+            result.add(futureStatefulSet);
+
+            if (deleteClaims) {
+                for (int i = 0; i < kafka.getReplicas(); i++) {
+                    Future<Void> f = Future.future();
+                    DeleteOperation.deletePersistentVolumeClaim(vertx, k8s.getKubernetesClient()).delete(namespace, kafka.getVolumeName() + "-" + kafka.getName() + "-" + i, f.completer());
+                    result.add(f);
+                }
+            }
+
+            return result;
+        }
+
+        @Override
+        public KafkaCluster getCluster(K8SUtils k8s, String namespace, String name) {
+            return KafkaCluster.fromStatefulSet(k8s, namespace, name);
+        }
+    };
+
+    @Override
+    protected Op<KafkaCluster> createOp() {
+        return create;
+    }
+
+    @Override
+    protected Op<KafkaCluster> deleteOp() {
+        return delete;
+    }
+
+    public void update(String namespace, String name, Handler<AsyncResult<Void>> handler) {
```

Comment:
Why doesn't `update` use the `Op<>` interface?

Reply:
`Op` is really the same thing as the `SimpleClusterOperation`, which was already the base class for creating and deleting cluster, but not for updating them. It might be possible to make the update operation fit with the pattern used for create and delete, but it's not completely similar: `create()` and `delete()` basically `join()` on a list of `Future`s, whereas the `Future.update()` uses `compose()`. But if you abstract the creation of the future, they all end up setting a handler on the resulting future. Let me look at this.

Reply:
But I'm not sure that whether the `create` / `delete` will be a join forever. For example for exposing Kafka to the outside - we would need to first create the services, collect some endpoint information once they are created (e.g. a loadbalancer address on Kubernetes) and only than create Kafka Statefulset with the endpoints being passed in as advertised listeners. So I think we will need to deal with this in any case, because the create work will not be only a simple join forever. It is like that now because it has still a lot of TODOs.

Reply:
I've refactored this (locally) so that `Op.futures()` (now called `CompositeOperation.composite()`)  returns the single combined Future, and the `execute()` method just sets a handler on that. This allows some flexibility in the future for non-join composites. This also means the `CompositeOperation` almost works for `update()` too. The reason it doesn't is because for update we need to pass a pair of objects between getCluster() and futures() -- the cluster and the diff. I guess we could create a `Pair` class to do this, but I've not done so yet. Wdyt?

Reply:
@tombentley are you referring to the `CompositeOperation<KafkaCluster>` which in the `update()` case should have the `KafkaCluster` but even the `ClusterDiffResult` ?

Reply:
That's right @ppatierno. I guess we could have the `getCluster()` always return a (`Cluster`, `ClusterDiffResult`)-pair and just let the diff result be `null` in the create and delete cases. That would conceptually still make sense.

Reply:
Yes it makes sense to me. Another possibility could be having a `getType` in the `CompositeOperation` interface in order to expose what kind of operation it's bringing (create, update, delete). Of course, it should expose a `getDiff` as well for returning the `ClusterDiffResult`. Finally, the `execute` in the `AbstractClusterOperations` should check the type from the composite operation to know if the `getDiff` call is needed and then behaving differently (which is something it needs to do in any case when the diff result is null or not).